lab/src/mapreduce/common_map.go

```diff
@@ -53,6 +55,29 @@ func doMap(
 	//
 	// Your code here (Part I).
 	//
+
+  	// read input file to buffer
+	file, _ := os.Open(inFile)
+	fileStat, _ := file.Stat()
+	buffer := make([]byte, fileStat.Size())
+	file.Read(buffer)
+	file.Close()
+
+  	// do map
+	result := mapF(inFile, string(buffer))
+
+  	// write result to intermediate files
+	for r := 0; r < nReduce; r++ {
+		interFileName := reduceName(jobName, mapTask, r)
+		file, _ := os.Create(interFileName)
+		enc := json.NewEncoder(file)
+		for _, kv := range result {
```

Comment:
A little improve here:
You are calling ihash for same key for multiple r several times. Rather than having this double loop(calling same api several times for same input) I think it is better to cache the result of same input API call

Reply:
Talked offline. Good idea!